examples/plot_metric_examples.py

```diff
@@ -0,0 +1,474 @@
+"""
+
+# License: BSD 3 clause
+# Authors: Bhargav Srinivasa Desikan <bhargavvader@gmail.com>
+#          William de Vazelhes <wdevazelhes@gmail.com>
+
+######################################################################
+# Imports
+# ^^^^^^^
+#
+
+from sklearn.manifold import TSNE
+
+import metric_learn
+import numpy as np
+from sklearn.datasets import make_classification, make_regression
+
+# visualisation imports
+import matplotlib.pyplot as plt
+np.random.seed(42)
+
+
+######################################################################
+# Loading our dataset and setting up plotting
+# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+#
+# We will be using a synthetic dataset to illustrate the plotting,
+# using the function `sklearn.datasets.make_classification` from
+# scikit-learn. The dataset will contain:
+#   - 100 points in 3 classes with 2 clusters per class
+#   - 5 features, among which 3 are informative (correlated with the class
+#     labels) and two are random noise with large magnitude
+
+X, Y = make_classification(n_samples=100, n_classes=3, n_clusters_per_class=2,
+                           n_informative=3, class_sep=4., n_features=5,
+                           n_redundant=0, shuffle=True,
+                           scale=[1, 1, 20, 20, 20])
+
+###########################################################################
+# Note that the dimensionality of the data is 5, so to plot the
+# transformed data in 2D, we will use the t-sne algorithm. (See
+# `sklearn.manifold.TSNE`).
+
+
+def plot_tsne(X, Y, colormap=plt.cm.Paired):
+    plt.figure(figsize=(8, 6))
+
+    # clean the figure
+    plt.clf()
+
+    tsne = TSNE()
+    X_embedded = tsne.fit_transform(X)
+    plt.scatter(X_embedded[:, 0], X_embedded[:, 1], c=Y, cmap=colormap)
+
+    plt.xticks(())
+    plt.yticks(())
+
+    plt.show()
+
+###################################
+# Let's now plot the dataset as is.
+
+
+plot_tsne(X, Y)
+
+#########################################################################
+# We can see that the classes appear mixed up: this is because t-sne
+# is based on preserving the original neighborhood of points in the embedding
+# space, but this original neighborhood is based on the euclidean
+# distance in the input space, in which the contribution of the noisy
+# features is high. So even if points from the same class are close to
+# each other in some subspace of the input space, this is not the case in the
+# total input space.
+#
+# Metric Learning
+# ^^^^^^^^^^^^^^^
+#
+# Why is Metric Learning useful? We can, with prior knowledge of which
+# points are supposed to be closer, figure out a better way to understand
+# distances between points. Especially in higher dimensions when Euclidean
+# distances are a poor way to measure distance, this becomes very useful.
+# 
+# Basically, we learn this distance:
+# :math:`D(x,y)=\sqrt{(x-y)\,M^{-1}(x-y)}`. And we learn this distance by
+# learning a Matrix :math:`M`, based on certain constraints.
```

Comment:
- there is repetition here "we learn this distance"
- add very quick explanation of what is meant by "constraint" (can be inspired from What is metric learning? page). For instance something like, "we learn the parameters :math:`M` of this distance to satisfy certain constraints on the distance between points, for example requiring that points of the same class are close together and points of different class are far away."

Reply:
Agreed, done